Add Kconfigs for `PSA_WANT` symbols

valeriosetti opened the pull request:
With the goal to move toward PSA crypto APIs (https://github.com/zephyrproject-rtos/zephyr/issues/43712) it is useful to have the possibility to define `PSA_WANT_xxx` features that are required. So far all of these symbols are automatically guessed in MbedTLS' header files by starting from `MBEDTLS_xxx` features. However there are some limitation to this:

- `PSA_WANT_xxx` symbols are not 1:1 mapped with `MBEDTLS_xxx` ones
- for some `PSA_WANT_xxx` there is no mapping with `MBEDTLS_xxx` (ex: `PSA_WANT_KEY_TYPE_DERIVE`)

For this reason this PR introduces `Kconfig` and an header file to support `PSA_WANT_xxx` symbols with a mechanism which mimics the already existing `Kconfig.tls-generic`/`config-tls-generic.h` combo. These new Kconfigs will be used whenever there is some crypto provider enabled, i.e. when `MBEDTLS_PSA_CRYPTO_C` or `BUILD_WITH_TFM` are set.

valeriosetti committed:
mbedtls: add configuration file for PSA features

Adding a Kconfig and a header file to control which PSA features
are to be used from MbedTLS:

- new kconfig symbols are placed in a separate header file
  (Kconfig.psa) and are guarded by MBEDTLS_PSA_CRYPTO_CLIENT. The
  reason for this is that TLS/X509 can either rely on PSA functions
  provided by MbedTLS (when MBEDTLS_PSA_CRYPTO_C is defined) or
  TFM (when BUILD_WITH_TFM is selected). Therefore we could
  not make these new Kconfigs depending on MBEDTLS_PSA_CRYPTO_C.

- by default all PSA symbols are enabled, but they can be
  disabled by respective Kconfigs in order to reduce the image
  size.

- the new header file (config-psa-generic.h) mimics what
  config-tls-generic.h does for MbedTLS builtin symbols: it
  enables a build symbol for every Kconfig one. The name is
  kept identical in the 2 cases (a part from the initial CONFIG_
  in order to simplify the mechanism).

- MBEDTLS_PSA_CRYPTO_CONFIG is finally enabled whenever there
  is any PSA crypto provider (either MBEDTLS_PSA_CRYPTO_C or
  TFM)

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: add the python script used to generate PSA Kconfigs

Since PSA has several different symbols defined and since it's
better to keep them aligned with the current MbedTLS release,
a python script is used to generate both:
- Kconfig.psa
- config-psa-generic.h

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
CI is green so this PR is ready for initial reviews.

valeriosetti committed:
mbedtls: default enable USE_PSA_CRYPTO when CRYPTO_C

Now that support for all PSA_WANT symbols is in place, we can
enable MBEDTLS_USE_PSA_CRYPTO when MBEDTLS_PSA_CRYPTO_C is
enabled as well.

Note: this commit also moves USE_PSA_CRYPTO out of CRYPTO_C
dependency in config-tls-generic.h because TLS/DTLS/X509 modules
of MbedTLS can rely on *any* implementation of PSA crypto APIs
not only the MbedTLS one. TFM is for example an alternative
to this.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: add new PSA_CRYPTO_CLIENT config

Using MBEDTLS_PSA_CRYPTO_CLIENT to guard all PSA_WANT symbols is
not completely correct because:

1. the prefix MBEDTLS suggests that it's something related to
   MbedTLS, while actually PSA APIs can be provided also
   by other implementations (ex: TFM)

2. there might applications which are willing to use PSA APIs
   without using MbedTLS at all. For example computing an hash
   can be as simple as writing psa_hash_compute() and, if the
   PSA function is provided thorugh TFM, then MbedTLS is not
   required at all

Therefore this commit:

- moves MBEDTLS_PSA_CRYPTO_CLIENT to Kconfig.tls-generic since
  that symbol belongs to MbedTLS

- adds a new symbol named PSA_CRYPTO_CLIENT as a generic way
  to state that there is "some" PSA crypto API provider

- let MBEDTLS_PSA_CRYPTO_CLIENT automatically select
  PSA_CRYPTO_CLIENT, since the former is an implementation of
  the latter.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

frkv commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -488,6 +488,7 @@ config MBEDTLS_PSA_CRYPTO_C
 
 config MBEDTLS_USE_PSA_CRYPTO
 	bool "Use PSA APIs instead of legacy MbedTLS when possible"
+	default y if MBEDTLS_PSA_CRYPTO_C
```
This should be MBEDTLS_PSA_CRYPTO_CLIENT

valeriosetti replied:
Correct, thanks! Fixed

valeriosetti committed:
mbedtls: add new source file from MbedTLS 3.6 release

Add "sha3.c" source file that was previously missing from
CMakeLists.txt and that has been added in MbedTLS 3.6 release.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

ithinuel reviewed (commented):
That is a lot of new config. I didn't check each individually but it looks alright.

IIUC `Kconfig.psa`and `config-psa.h` are generated by  `create_psa_files.py`.

- Is it expected to be executed manually every now and then?
- Is there any CI setup that can run it an check it does not generate any change?
  It would be nice to have this to detect changes in compatibility/naming.

alxelax commented on tests/bsim/bluetooth/mesh/overlay_psa.conf:
```diff
@@ -5,3 +5,4 @@ CONFIG_BT_GATT_CACHING=n
 
 # Enable mbedTLS PSA as a crypto backend
 CONFIG_BT_MESH_USES_MBEDTLS_PSA=y
+CONFIG_PSA_CRYPTO_ENABLE_ALL=y
```
I do not think I agree that BLE Mesh requires all crypto PSA features. There is list of features that Mesh uses, those are `selected` during `BT_MESH_USES_MBEDTLS_PSA` activation.

alxelax replied:
There is one more thing. As a Bluetooth Mesh user, I expect that when I chose crypto backend, this is sufficient to build my application. This extra option seems like hidden knowledge that isn't obvious to people not known mbedtls-psa configuration system.

tomi-font replied:
> I do not think I agree that BLE Mesh requires all crypto PSA features. There is list of features that Mesh uses, those are `selected` during `BT_MESH_USES_MBEDTLS_PSA` activation.

So far the PSA configuration has been defined by MbedTLS's default configuration file, `psa/crypto_config.h`, which unconditionally does `#define PSA_WANT_[...] 1` for every symbol. In this regard this PR does not change anything. It introduces the configuration of `PSA_WANT` symbols for future work to build on top of this. Also note that when TF-M is in use, currently the actual crypto features available are defined by what TF-M profile is used.

However I am not certain how this fares when TF-M is not in use. It could be that with MbedTLS's adjustment of legacy symbols from PSA ones this would inflate the final code size for nothing.

> There is one more thing. As a Bluetooth Mesh user, I expect that when I chose crypto backend, this is sufficient to build my application. This extra option seems like hidden knowledge that isn't obvious to people not known mbedtls-psa configuration system.

Good point. Could we instead `select PSA_CRYPTO_ENABLE_ALL` @valeriosetti? (And you may also want to add it under `config BT_MESH_USES_TFM_PSA`, actually.) Or then, because of the first point above, select only needed PSA symbols. Or simply stick to the legacy ones and `MBEDTLS_USE_PSA_CRYPTO` if possible?

valeriosetti replied:
As @alxelax suggested `BT_MESH_USES_MBEDTLS_PSA` selects required MbedTLS features, so I can try to replace legacy `MBEDTLS_` symbols with proper `PSA_WANT_` ones. This should fix both comments.
I'll give this a try

valeriosetti replied:
@alxelax I reverted the previous change and I've added https://github.com/zephyrproject-rtos/zephyr/pull/72243/commits/513fb7d339a697eae7b7c1920123f0e909c927a9 which should match both your requirements. Wdyt?

alxelax replied:
Yep, now it looks much better. Seems it missed some option(s) since devices cannot generate public key in bsim tests.

valeriosetti replied:
Yes, I should have fixed both EC key generation issue and a missing  `PSA_WANT_ALG_ECB_NO_PADDING`. Hopefully bsim will be fine now.

tomi-font commented on modules/mbedtls/create_psa_files.py:
```diff
@@ -0,0 +1,89 @@
+if PSA_CRYPTO_CLIENT
+
+config PSA_CRYPTO_ENABLE_ALL
+	bool
+	prompt "All PSA features enabled"
```
```suggestion
	bool "All PSA features"
```

valeriosetti replied:
All prompts and description for `PSA_CRYPTO_ENABLE_ALL` should be fixed now

tomi-font commented:
> Is it expected to be executed manually every now and then?

It would need to be run every time we 1. want to make a change to something in `Kconfig.psa` or 2. update MbedTLS.

> Is there any CI setup that can run it an check it does not generate any change?
It would be nice to have this to detect changes in compatibility/naming.

I have also thought about the possibility of having an automatic check to ensure that the committed header file is exactly what is generated by the script. Could be nice if it's not too involved to do.

valeriosetti committed:
bluetooth: mesh: update BT_MESH_USES_MBEDTLS_PSA selected symbols

Instead of selecting legacy MBEDTLS symbols, use corresponding
PSA_WANT ones when possible (note: some legacy symbols do not
have a PSA correspondance).

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on modules/mbedtls/Kconfig.tls-generic:
```diff
@@ -105,7 +105,7 @@ config MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
 config MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED
 	bool "ECDH-ECDSA based ciphersuite modes"
-	depends on MBEDTLS_ECDH_C && MBEDTLS_ECDSA_C
+	depends on (MBEDTLS_ECDH_C && MBEDTLS_ECDSA_C) || (PSA_WANT_ALG_ECDH && PSA_WANT_ALG_ECDSA)
```
Are we gonna have other MbedTLS Kconfig options start depending on PSA ones? :thinking:

frkv replied:
Would it be more convenient (at least for the reader) to try to emulate the "CAN"-concept from Mbed TLS as Kconfigs?

Like `MBEDTLS_CAN_ECDH`
https://github.com/Mbed-TLS/mbedtls/blob/2ca6c285a0dd3f33982dd57299012dacab1ff206/include/mbedtls/config_adjust_legacy_crypto.h#L309C9-L309C25

Or more distinct: `MBEDTLS_PK_CAN_ECDSA_SOME`
https://github.com/Mbed-TLS/mbedtls/blob/2ca6c285a0dd3f33982dd57299012dacab1ff206/include/mbedtls/config_adjust_legacy_crypto.h#L333C9-L333C34

There are more than just `PSA_WANTS_ALG_XXXX` that can be used as a signal, since you can choose to use or not use PSA crypto with `MBEDTLS_USE_PSA_CRYPTO`. Arguably we of course prefer the `MBEDTLS_USE_PSA_CRYPTO` and PSA crypto will at some point likely be the only option, but it is still slightly more readable no not have to rehash both legacy and PSA configs for every ciphersuite

When I start thinking about it, there is an option to have a stable configuration follow the pattern `MBEDTLS_CAN_ECDH_ECDSA_KEY_EXCHANGE` for every relevant key exchange/cipher-suite

We have made a similar concept in nRF Connect SDK

valeriosetti replied:
TBH in my first attempt I was about to remove those "depends on" because this is totally duplicating checks that MbedTLS already performs in `check_config.h` and, in case something is missing, it already prints a clear message of what's missing and where to look at.
Honestly I'm not a huge fan of duplicating also the `MBEDTLS_CAN_ECDH` and other MbedTLS internal symbols, because this adds extra burden of work every time Zephyr updates the MbedTLS library because you need to update all these checks again and again.
In my opinion I would follow the same path I've introduced for the `PSA_WANT` symbols: just add 1 Kconfig symbol for every MbedTLS build symbol, but no extra logic. Then if something is not correct `check_config.h` will error out some message. No extra work for Zephyr. However this proposed approach imposes quite some change to the current `Kconfig.tls-generic" file so I would keep this for a follow up PR in case we find an agreement (@tomi-font ).
Wdyt?

tomi-font replied:
Agreed to keep further significant work for a follow-up PR. And at first thought it seems senseful to get rid of `depends on` statements if those checks are performed by MbedTLS itself.

frkv replied:
This is for sure follow-up work. This only makes sense if we have stable definitions in Mbed TLS distribution. Arguably PSA_WANT_ALG_XXXX should be sufficient for most cases, but then this can only happen when mbedtls_ prefixed crypto toolbox APIs are removed from the system (e.g. in downstream Zephyr or in upstream Mbed TLS)

tomi-font commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1350,15 +1350,17 @@ config BT_MESH_USES_MBEDTLS_PSA
-	select MBEDTLS_MAC_CMAC_ENABLED
-	select MBEDTLS_CIPHER_AES_ENABLED
+	select PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC
+	# Required otherwise there are missing symbols for PK helpers at link time.
```
What line(s) is this comment addressing?

valeriosetti replied:
The line below. It enables RSA keys support which might look odd since we're only using ECDH and ECDSA, so I thought adding a comment there could help. Doesn't it?

tomi-font replied:
I see. I didn't pay much attention to what is (meant to be) used compared to what is enabled. Others can likely judge better. The comment is fine and makes sense.

frkv replied:
```suggestion
	# Required, otherwise there are missing symbols for PK helpers at link time.
```
I think a comma is needed

valeriosetti commented:
> > Is there any CI setup that can run it an check it does not generate any change?
> > It would be nice to have this to detect changes in compatibility/naming.
> 
> I have also thought about the possibility of having an automatic check to ensure that the committed header file is exactly what is generated by the script. Could be nice if it's not too involved to do.

I totally agree on this and I will investigate on it.

fabiobaltieri commented:
> I totally agree on this and I will investigate on it.

Since the script is in modules there's no current workflow that looks for pytest in there. I think you have two options:

- you could write a pytest test and add the path to https://github.com/zephyrproject-rtos/zephyr/blob/main/.github/workflows/scripts_tests.yml#L71 (then you can probably mock the write function and make sure it's called with what you expect, you may have to refactor the script a bit to make it testable), or
- just add a test to validate against a golden sample in an actual build only test, see https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/misc/check_init_priorities for an example, this may be easier if you already have a build test that runs this script, just add the validation step on it or something

tomi-font reviewed (dismissed):
ping @ceolin @ithinuel @d3zd3z

valeriosetti committed:
test: mbedtls: add custom target to check MbedTLS auto-generated files

This commit adds a custom target to tests/crypto/mbedtls in
order to verify that MbedTLS auto-generated files are up to
date.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on modules/mbedtls/Kconfig.psa:
```diff
@@ -1,14 +1,354 @@
 # SPDX-License-Identifier: Apache-2.0
 
-config MBEDTLS_PSA_CRYPTO_CLIENT
+# This file was automatically generated by create_psa_files.py.
+# Do not edit it manually.
```
While implementing this I realized that the only relevant info to be added here is probably the file that is used as reference to generate the other ones. The reason for skipping version/revision is that MbedTLS can change (likely be upgraded over time), but what we really care of here is the `crypto_config.h` file. If that file does not change, then PSA is not introducing new symbols so Zephyr does not need to adapt.

valeriosetti replied:
This should be fixed now

de-nordic replied:
I guess ok.

valeriosetti commented:
> > I totally agree on this and I will investigate on it.
> 
> Since the script is in modules there's no current workflow that looks for pytest in there. I think you have two options:
> 
>     * you could write a pytest test and add the path to https://github.com/zephyrproject-rtos/zephyr/blob/main/.github/workflows/scripts_tests.yml#L71 (then you can probably mock the write function and make sure it's called with what you expect, you may have to refactor the script a bit to make it testable), or
> 
>     * just add a test to validate against a golden sample in an actual build only test, see https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/misc/check_init_priorities for an example, this may be easier if you already have a build test that runs this script, just add the validation step on it or something

@fabiobaltieri I think I need a quick help on this. I modified the python script so that it's also able to just check if the already existing files are up to date or not and I also added a check in `twister.yaml`. However I don't see this test appearing on the CI. What am I missing?

fabiobaltieri commented on .github/workflows/twister.yaml:
```diff
@@ -206,6 +206,11 @@ jobs:
           echo "github.base_ref: ${{ github.base_ref }}"
           echo "github.ref_name: ${{ github.ref_name }}"
 
+      - name: Check MbedTLS auto-generated files
+        run: |
+          echo "Check MbedTLS auto-generated files"
+          ./modules/mbedtls/create_psa_files.py --check
+
```
This should not really be in twister.yaml, it could go in a different workflow but since it's implemented as a check option rather than an actualy pytest I don't there's a good place to put it.

How about instead of trying to run it directly from a workflow you add the check to one of the existing test or samples? (like the other example I pointed out in my comment). Maybe this could live as part of the tfm_integration test

valeriosetti replied:
>  it could go in a different workflow

Initially I added this check in `scripts_test.yaml` as you suggested, but that workfile is executed on the Zephyr repo itself while my script requires the MbedTLS repo to be present as well in order to work properly (i.e. `west init` must be issued). That's why I moved it from `scripts_test.yaml` to `twister.yaml`

> How about instead of trying to run it directly from a workflow you add the check to one of the existing test or samples?

I will try this option

valeriosetti replied:
Turns out that I overlooked the 2nd option you wrote [previously](https://github.com/zephyrproject-rtos/zephyr/pull/72243#issuecomment-2111907143), but it was actually the easiest way to achieve what I was meant to do here. This should now be fixed in https://github.com/zephyrproject-rtos/zephyr/commit/37769f4f74fc5754a4e8f75abd9131eea68ed8cb.
Thanks again!

frkv commented on modules/mbedtls/Kconfig.psa:
```diff
@@ -1,14 +1,355 @@
+if PSA_CRYPTO_CLIENT
+
+config PSA_CRYPTO_ENABLE_ALL
+	bool "All PSA features"
```
```suggestion
	bool "Enable all PSA crypto features"
```

Minor NIT

valeriosetti replied:
I think I had issues where the CI was complaining about prompt starting with "Enable", so that's why I removed that part. I can add the "crypto" though

frkv reviewed (approved):
LGTM (minor NITS and assuming correct placement of the check-script is resolved)

valeriosetti commented:
I should have addressed all the comments now and the check on auto-generated files is active (implementation is [here](https://github.com/zephyrproject-rtos/zephyr/pull/72243/commits/22e7ac3922c2aef8b737af7e96cd3e50111038d1))

tomi-font reviewed (approved):
Very nice job!

Asking @fabiobaltieri as I couldn't find anything relating to the newly added check in the CI logs. Does it seem okay to you now, and will it be run in CI both if something in `modules/mbedtls` is modified and also when MbedTLS's manifest revision is changed?

valeriosetti commented:
Would it possible to merge this PR now that we have a good amount of approvals, please? This helps unlocking other depending PRs

aescolar commented:
@valeriosetti Given that this can cause problems for some apps, should we have an entry in the migration guide?

valeriosetti commented:
Yes sure! I can work on that. I was also requested to review the crypto documentation changes in https://github.com/zephyrproject-rtos/zephyr/pull/73241
